### PR TITLE
refactor: omnichannel's on-hold feature

### PR DESCRIPTION
**Pull request by murtaza98**
## Issue(s)

https://rocketchat.atlassian.net/browse/OC-529
## Further comments

- [x] deprecate meteor method to resume on-hold chat's and create new API for this action

### CODE REVIEW

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import { Meteor } from 'meteor/meteor';
+import type { IMessage, IOmnichannelRoom, IUser } from '@rocket.chat/core-typings';
+import { isOmnichannelRoom } from '@rocket.chat/core-typings';
+import { LivechatRooms, Subscriptions, Messages, LivechatInquiry } from '@rocket.chat/models';
+
+import { RoutingManager } from '../../../../../app/livechat/server/lib/RoutingManager';
+import { callbacks } from '../../../../../lib/callbacks';
+import { ohLogger as logger } from './logger';
+
+class OnHoldHelperClass {
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:10]: Let's move this to be a service, under ee/server/services

File: apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts [hunk 2]
```diff
@@ -1,21 +1,21 @@
...
 	const dispatchToastMessage = useToastMessageDispatch();
 
 	const queryClient = useQueryClient();
 
 	return useMutation(
-		async (rid) => {
+		async (roomId) => {
 			try {
-				await resumeChatOnHold(rid, { clientAction: true });
+				await resumeChatOnHold({ roomId });
```
Comment (KevLehman) [inline apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts:18]: what about clientAction? 👀

File: apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts [hunk 3]
```diff
@@ -1,21 +1,21 @@
...
 			} catch (error) {
 				console.error(error);
```
Comment (KevLehman) [inline apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts:20]: Not part of the PR, but, do we need this console.log?

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 4]
```diff
@@ -51,7 +52,52 @@ API.v1.addRoute(
...
+	'livechat/room.resumeOnHold',
+	{ authRequired: true, permissionsRequired: ['view-l-room'], validateParams: isLivechatRoomResumeOnHoldProps },
+	{
+		async post() {
+			const { roomId } = this.bodyParams;
+			if (!roomId || roomId.trim() === '') {
+				throw new Error('invalid-param');
+			}
+
+			const room = await LivechatRooms.findOneById(roomId);
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:72]: do we need whole room info?

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+const resumeOnHoldCommentAndUser = async (room: IOmnichannelRoom): Promise<{ comment: string; resumedBy: IUser }> => {
+	const {
+		v: { _id: visitorId },
+		_id: rid,
+	} = room;
+	const visitor = await LivechatVisitors.findOneById<Pick<ILivechatVisitor, 'name' | 'username'>>(visitorId, {
+		projection: { name: 1, username: 1 },
+	});
+	if (!visitor) {
+		throw new Error(`[afterSaveMessage] Visitor Not found for room ${rid} while trying to resume on hold`);
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts:19]: Can we replace this with a normal error + a detailed log?

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts (hunk 1 above)
Comment (KevLehman) [reply apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:10]: This could be a "omnichannel-ee" service that will hold ee features for omni 🤗

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+			name: onHoldBy.name || '',
+		};
+
+		await Promise.all([
+			LivechatRooms.setOnHoldByRoomId(roomId),
+			Subscriptions.setOnHoldByRoomId(roomId),
+			Messages.createOnHoldHistoryWithRoomIdMessageAndUser(roomId, onHoldByUser, comment, 'on-hold'),
+		]);
+
+		Meteor.defer(() => {
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:38]: Do we need the meteor.defer? can't we move this to the end of func?

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+		};
+
+		await Promise.all([
+			LivechatRooms.unsetOnHoldByRoomId(roomId),
+			Subscriptions.unsetOnHoldByRoomId(roomId),
+			Messages.createOnHoldHistoryWithRoomIdMessageAndUser(roomId, resumeByUser, comment, 'resume-onHold'),
+		]);
+
+		const updatedRoom = await LivechatRooms.findOneById(roomId, {});
+		if (updatedRoom) {
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:95]: Same here

File: apps/meteor/ee/app/livechat-enterprise/server/methods/resumeOnHold.ts [hunk 8]
```diff
@@ -60,7 +65,6 @@ Meteor.methods({
 		const comment = await resolveOnHoldCommentInfo(options, room, onHoldChatResumedBy);
 		(Messages as any).createOnHoldResumedHistoryWithRoomIdMessageAndUser(roomId, comment, onHoldChatResumedBy);
 
-		const updatedRoom = LivechatRooms.findOneById(roomId);
-		updatedRoom && Meteor.defer(() => callbacks.run('livechat:afterOnHoldChatResumed', updatedRoom));
+		Meteor.defer(() => callbacks.run('livechat:afterOnHoldChatResumed', room));
```
Comment (KevLehman) [inline apps/meteor/ee/app/livechat-enterprise/server/methods/resumeOnHold.ts:68]: Won't have any problems by using an outdated room instead of the updated one?

File: apps/meteor/ee/app/models/server/models/Messages.js [hunk 9]
```diff
@@ -18,6 +19,7 @@ Messages.prototype.createOnHoldHistoryWithRoomIdMessageAndUser = function (roomI
 	return record;
 };
 
+// TODO: rm this
```
Comment (KevLehman) [inline apps/meteor/ee/app/models/server/models/Messages.js:22]: 🤔

**Review by KevLehman — CHANGES_REQUESTED** [review]

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts (hunk 1 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:10]: Let me try this. I wanted to do this before but since this method had some meteor code in the form of callbacks, I went with this approach.

File: apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts (hunk 3 above)
Comment (murtaza98) [reply apps/meteor/client/views/room/components/body/composer/ComposerOmnichannel/hooks/useResumeChatOnHoldMutation.ts:20]: Not needed, the dispatch method has an internal logger which logs error afaik. So let me remove it

File: apps/meteor/ee/app/models/server/models/Messages.js (hunk 9 above)
Comment (murtaza98) [reply apps/meteor/ee/app/models/server/models/Messages.js:22]: oops! forgot to remove these

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts (hunk 1 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:10]: Moved it to service and everything is working fine :partying_face:

File: apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts (hunk 6 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/lib/OnHoldHelper.ts:38]: Apparently it works without defer, so I'm gonna remove it since it's from Meteor and we're moving this to a service which is known to sometimes cause issues with meteor

File: apps/meteor/ee/app/livechat-enterprise/server/methods/resumeOnHold.ts (hunk 8 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/methods/resumeOnHold.ts:68]: I've modified the callbacks to work without the updated room data. So we can save this DB call here :tada:

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts (hunk 4 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts]: nopes, applied more restrictive typing

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 10]
```diff
@@ -1,57 +1,108 @@
...
 
-			const user = Meteor.user();
-			if (!user) {
-				return API.v1.failure('Invalid user');
+			if (!room.servedBy) {
+				throw new Error('error-unserved-rooms-cannot-be-placed-onhold');
 			}
 
-			const subscription = Subscriptions.findOneByRoomIdAndUserId(roomId, user._id, { _id: 1 });
+			const subscription = Subscriptions.findOneByRoomIdAndUserId(roomId, this.userId, { _id: 1 });
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:49]: is this part needing to change to `this.userId`?

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 11]
```diff
@@ -1,57 +1,108 @@
...
 			if (!subscription && !hasPermission(this.userId, 'on-hold-others-livechat-room')) {
-				return API.v1.failure('Not authorized');
+				throw new Error('Not_authorized');
 			}
 
-			const onHoldBy = { _id: user._id, username: user.username, name: (user as any).name };
+			const onHoldBy = { _id: this.userId, username: this.user.username, name: this.user.name };
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:54]: this one too, same issue, the `this.` changes

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 12]
```diff
@@ -1,57 +1,108 @@
...
+
+			if (!room.onHold) {
+				throw new Error('error-room-not-on-hold');
+			}
+
+			if (!room.open) {
+				throw new Error('This_conversation_is_already_closed');
+			}
+
+			const subscription = Subscriptions.findOneByRoomIdAndUserId(roomId, this.userId, { _id: 1 });
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:94]: and this

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 13]
```diff
@@ -1,57 +1,108 @@
...
+			if (!subscription && !hasPermission(this.userId, 'on-hold-others-livechat-room')) {
+				throw new Error('Not_authorized');
+			}
+
+			const { name, username, _id: userId } = this.user;
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:99]: and this

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 14]
```diff
@@ -1,57 +1,108 @@
...
+			const onHoldBy = { _id: userId, username, name };
+			const comment = TAPi18n.__('Omnichannel_on_hold_chat_resumed_manually', {
+				user: onHoldBy.name || `@${onHoldBy.username}`,
+			});
+
+			await OmnichannelEEService.resumeRoomOnHold(room, comment, this.user, true);
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:105]: and this

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/afterOnHoldChatResumed.ts [hunk 15]
```diff
@@ -1,20 +1,26 @@
...
-import { LivechatEnterprise } from '../lib/LivechatEnterprise';
+import { AutoCloseOnHoldScheduler } from '../lib/AutoCloseOnHoldScheduler';
 import { cbLogger } from '../lib/logger';
 
-const handleAfterOnHoldChatResumed = async (room: any): Promise<void> => {
-	if (!room || !room._id || !room.onHold) {
-		cbLogger.debug('Skipping callback. No room provided or room is not on hold');
-		return;
+const handleAfterOnHoldChatResumed = async (room: Pick<IOmnichannelRoom, '_id'>): Promise<Pick<IOmnichannelRoom, '_id'>> => {
+	if (!room?._id) {
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/hooks/afterOnHoldChatResumed.ts:8]: shouldn't you add `!room.onHold` back?

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+		callbackLogger.error(`[afterSaveMessage] Visitor Not found for room ${rid} while trying to resume on hold`);
+		throw new Error('Visitor not found while trying to resume on hold');
+	}
+
+	const guest = visitor.name || visitor.username;
+
+	const resumeChatComment = TAPi18n.__('Omnichannel_on_hold_chat_automatically', { guest });
+
+	const resumedBy = await Users.findOneById('rocket.cat');
+	if (!resumedBy) {
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts:28]: can rocket.cat not exist?

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+};
+
+const handleAfterSaveMessage = async (message: IMessage, room: IRoom) => {
+	if (isEditedMessage(message) || message.t || !isOmnichannelRoom(room)) {
+		return message;
+	}
+
+	const { _id: rid, v: roomVisitor } = room;
+
+	if (!roomVisitor?.token) {
```
Comment (cauefcr) [inline apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts:43]: token should not be used as ID

File: apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts [hunk 18]
```diff
@@ -122,6 +158,97 @@ import { IS_EE } from '../../../e2e/config/constants';
...
+		});
+		it('should resume chat automatically if visitor sent a message', async () => {
+			const { room, visitor } = await startANewLivechatRoomAndTakeIt();
+
+			await sendAgentMessage(room._id);
+			await placeRoomOnHold(room._id);
+			await sendMessage(room._id, 'test', visitor.token);
+
+			// wait for the room to be resumed since that logic is within callbacks
+			await sleep(1000);
```
Comment (cauefcr) [inline apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts:248]: can we decrease the sleep time?

**Review by cauefcr — CHANGES_REQUESTED** [review]
left some comments

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts (hunk 10 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:49]: Although its not needed, I still would like to keep this since this is the recommended way to get the logged in user. If you look at the previous code, it had a Meteor dependency which we're trying to avoid as much as possible.

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts (hunk 11 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:54]: https://github.com/RocketChat/Rocket.Chat/pull/28306/files#r1136560792

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts [hunk 19]
```diff
@@ -1,57 +1,108 @@
...
 			}
 
-			const onHoldBy = { _id: user._id, username: user.username, name: (user as any).name };
+			const onHoldBy = { _id: this.userId, username: this.user.username, name: this.user.name };
 			const comment = TAPi18n.__('Omnichannel_On_Hold_manually', {
 				user: onHoldBy.name || `@${onHoldBy.username}`,
 			});
 
-			await LivechatEnterprise.placeRoomOnHold(room, comment, onHoldBy);
+			await OmnichannelEEService.placeRoomOnHold(room, comment, this.user);
```
[the remark this answers is not in the document]
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:59]: https://github.com/RocketChat/Rocket.Chat/pull/28306/files#r1136560792

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts (hunk 12 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:94]: https://github.com/RocketChat/Rocket.Chat/pull/28306/files#r1136560792

File: apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts (hunk 14 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/api/rooms.ts:105]: https://github.com/RocketChat/Rocket.Chat/pull/28306/files#r1136560792

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/afterOnHoldChatResumed.ts (hunk 15 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/hooks/afterOnHoldChatResumed.ts:8]: No, its not required now since this callback will only get called when a room is resumed from on-hold, since its `livechat:afterOnHoldChatResumed` :sweat_smile:

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts (hunk 16 above)
Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts:28]: Then they are screwed :joy: A lot of our features like omnichannel-transcript, onhold, etc are built around using this `rocket.cat` user as a system user. So if some workspace doesn't have this user, then we should consider it as mis-configured and ask them to add this user back

File: apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts (hunk 17 above)
Comment (murtaza98) [inline apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts:43]:
```suggestion
	if (!roomVisitor?._id) {
```

Comment (murtaza98) [reply apps/meteor/ee/app/livechat-enterprise/server/hooks/resumeOnHold.ts]: Good catch!, let me change that

File: apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts (hunk 18 above)
Comment (murtaza98) [reply apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts:248]: Sure, I can give it a try.

Comment (murtaza98) [inline apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts:248]:
```suggestion
			await sleep(500);
```

Comment (murtaza98) [reply apps/meteor/tests/end-to-end/api/livechat/18-rooms-ee.ts]: Looks like 500 ms also works!

**Review by cauefcr — APPROVED** [review]
